feat: Scrubbing of UTF-16 strings in minidumps

flub opened the pull request:
This first scrubs UTF-8 from the binary blobs, than scrubs UTF-16 from the remainder.

The functionality is there and mostly in the right order and shape.

- [x] Refactor to avoid matches vector
- [x] Maybe implement `StringMods` for UTF-8/`str`
- [x] Maybe merge `apply_regex_to_*_slices`
- [x] Merge master
- [x] Add changelog

Commit:
Basic UTF-16 string extraction from random memory

This doens't yet do anything with the strings, they are also mostly
unicode garbage but we don't care as they also contain the readable
parts which is what we want to match on.

Commit:
implement scrubing of utf16 bytes

This is a rough first scetch.

Commit:
Add unittests, fix bug to count u16 as 2 bytes

Commit:
Merge branch 'master' into feat/utf16-scrub

Commit:
Fixup for simplified configs in master

Commit:
Add a WStr UTF-16LE type

Commit:
Add a safe constructor for WStr

Commit:
Move wstring to it's own crate

Commit:
Clean up the public interface

This now compiles without warnings.

Commit:
Make clippy happy

Commit:
Fixup docs

Commit:
Rename to relay-wstring for convenience

Commit:
Split off SliceIndex to it's own module

Commit:
Back to an iterator, this is much nicer

Commit:
Make this work again, using iteration and WStr

This is starting to look nicer.

Commit:
Some clippy fixes

Commit:
Break out scrubbing of matches to a fuction

This gets rid of another vector doing things inline in an iterator instead.

untitaker commented on relay-general/src/pii/attachments.rs:
```diff
@@ -110,6 +114,259 @@ fn apply_regex_to_bytes(
+fn scrub_match(all_text: &str, re_match: Match, redaction: &Redaction, all_encoded: &mut WStr) {
+    let mut segment_char_start = 0;
+    let mut segment_char_end = 0;
+    for (cur_char_offset, (cur_byte_offset, _cur_char)) in all_text.char_indices().enumerate() {
```
Can't this be done in one big `zip` without all the intermediate variables?

flub replied:
yep, done

untitaker commented on relay-general/src/pii/attachments.rs:
```diff
@@ -110,6 +114,259 @@ fn apply_regex_to_bytes(
+                self.offset += unprocessed_offset;
+                if decoded.len() > 2 {
+                    return Some(MutSegment {
+                        raw: unsafe { std::mem::transmute(&mut self.data[start..end]) },
```
Are we transmuting only the lifetimes here? In any case i'd explicitly specify the type parameters of transmute to make this more readable. Otherwise random stuff is inferred if the struct def is changed.

untitaker replied:
Also I think i'd prefer making this a function that takes a callback. instead of this `loop { }` and unsafe usage.

flub replied:
i made the transmute explicit.

I prefer the iterator over a callback i think.  But the loop would sadly stay though, it's a result of skipping over strings that are only 1 or 2 characters long.  I think it could be a reasonable call to not skip those, our regexes would either match or not anyway and then the loop can go.

flub replied:
So this loop is still an interesting design decision.  I think we should remove it and try to match regexes even if we only have 1 or 2 characters, in normal cases the regex would just not match.  In the cases where it would match it is a rather invisible bit of weird behaviour that we'd skip over it.

Plus this code would be simpler and safer :)

flub replied:
keep it, and make it 5

untitaker commented on relay-general/src/pii/attachments.rs:
```diff
@@ -110,6 +114,259 @@ fn apply_regex_to_bytes(
+        // WStr could be invalid, but after we called .fill_content() it will be valid.
+        let remainder_bytes = &mut self.as_bytes_mut()[offset..];
+        let remainder = unsafe { WStr::from_utf16le_unchecked_mut(remainder_bytes) };
+        remainder.fill_content(padding)
```
you're specifying that constructing wstr with invalid utf16 is UB but you're also assuming that `fill_content` is able to deal with a wstr that is partially invalid, a state that should not exist. There is not much in the way of somebody modifying either `from_utf16le_unchecked_mut` or `fill_content` to segfault or panic on invalid input each.

I think that a method called like `from_*_unchecked` is typically used to bypass expensive validation of input that is already valid, not to construct types in broken states. I would not construct a `String` with broken input either, even if I just call the most trivial method on it.

Perhaps you want `fill_content` to be a static method that takes bytes by value and returns wstr?

flub replied:
good catch, thanks! simplified and avoided using UB

untitaker commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,509 @@
+        let u = u16::from_le_bytes(buf);
+
+        if u < 0xD800 || 0xDFFF < u {
+            Some(unsafe { std::char::from_u32_unchecked(u as u32) })
```
Can't we use `from_u32` and catch the error? `from_u32` seems to almost have the same condition as you write here

```rust
std::char::from_u32(u as u32).unwrap_or_else(|| {
    // prev else branch here
})
```

flub replied:
hmm, i thought this was a neat idea, but after trying it i actually dislike it.  i think it's a case of "explicit is better than implicit" and since we do the decoder math in the else branch already to me it's clearer with the current if branch

untitaker replied:
can we do `from_u32().unwrap()` then? It should be optimized away

flub replied:
@untitaker argued that he prefers panics (unwrap) over unsafe so i followed this suggestion anyway.  there's one more if branch and an unwrap now.  i'm guessing this doesn't really matter performance wise

jan-auer replied:
Hm, I disagree with this. This is very central code in hot loops. If we can statically infer that this doesn't fail, then let's use unsafe over panics, or find a way to avoid the unsafety to begin with. This is also the approach the standard library is taking. 

In any case, do add comments explaining why this is infallible.

untitaker replied:
avoiding the unsafety was my first suggestion

flub replied:
So shall I revert this back to the unsafe version, perhaps with some more comments?

untitaker replied:
I think it's fine now

untitaker commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,509 @@
+    #[test]
+    fn test_wstr_is_ascii() {
+        let b = b"h\x00e\x00l\x00l\x00o\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
```
Can't we use the safe version of this with unwrap instead, makes it easier to audit the code

flub replied:
good call.  most these tests were written before the safe version existsed, hence why they looked like this.

untitaker reviewed (changes requested):
preliminary review of unsafe code

Commit:
Do not needlessly use unsafe in tests

Makes the tests easier to read.  Most of these tests were written
before the safe version existed...

Commit:
Explicit lifetimes transmute

flub commented on relay-general/src/pii/attachments.rs:
```diff
@@ -399,4 +719,101 @@ mod tests {
+        let mut data = Vec::from(&b"\xd8\xd8\xd8h\x00e\x00l\x00l\x00o\x00\xd8\xd8"[..]);
+        let mut iter = MutSegmentIter::new(&mut data, *UTF_16LE);
+
+        // Off-by-one is devastating, nearly everything is valid unicode.
+        let segment = iter.next().unwrap();
+        assert_eq!(segment.decoded, "棘攀氀氀漀");
```
This test is here purely to demonstrate the limitations btw.

Commit:
Trade off unsafe for more if statements and unwrap

Commit:
Implement FusedIterator for char iters

They already are.

Commit:
Force padding characters to be encoded to a single code

Changing these methods to have this restriction is makes them easier
to be correct, and it is all we need.  This also avoid using undefined
behaviour in WStr.

Commit:
Use zip to reduce iteractions some more

Also, break when we're done.

Commit:
Remove obsolete comment

Commit:
A bit more restructuring to make utf-8 code re-use easier

Commit:
Code re-use with utf-8 implementation

This doesn't actually change the behaviour of UTF-8 at all, despite
the more complicated things I envisaged before.  It's still possible
to do if we like though.

flub commented on relay-general/src/pii/attachments.rs:
```diff
@@ -71,43 +65,283 @@ fn apply_regex_to_bytes(
+    type Error = Utf16Error;
+
+    fn fill_content(&mut self, fill_char: char) {
+        let size = std::mem::size_of::<u16>();
```
I do this in a few functions.  However `std::mem::size_of` is a `const fn` so this is probably just a no-op?  In that case I'd prefer to write`std::mem::size_of::<u16>()` everywhere instead.  Unless I'm alone in this.

untitaker commented on relay-general/src/pii/attachments.rs:
```diff
@@ -329,7 +316,11 @@ impl<'a> Iterator for MutSegmentIter<'a> {
                     return Some(MutSegment {
-                        raw: unsafe { std::mem::transmute(&mut self.data[start..end]) },
+                        raw: unsafe {
+                            std::mem::transmute::<&'_ mut [u8], &'_ mut [u8]>(
```
cant we use `&'a` instead of `&'_`? not sure

flub replied:
I tried that but then it just sees the same as without transmute and refuses

untitaker commented on relay-wstring/src/lib.rs:
```diff
@@ -257,11 +256,13 @@ impl<'a> Iterator for WStrChars<'a> {
             let c = (((u - 0xD800) as u32) << 10 | (u2 - 0xDC00) as u32) + 0x1_0000;
-            Some(unsafe { std::char::from_u32_unchecked(c) })
-        }
+            Some(std::char::from_u32(c).unwrap())
```
Ah I overlooked this, that's a bummer

jan-auer replied:
@untitaker What's the action item here?

I'm curious if we can avoid this `unwrap`, potentially using [`std::hint::unreachable_unchecked`](https://doc.rust-lang.org/stable/std/hint/fn.unreachable_unchecked.html). Note that this differs from the `expect` above, as this instance here is indeed unreachable.

untitaker replied:
ok if you want to be unsafe you can use `from_u32_unchecked` like before. I think this is fine as-is

untitaker commented on relay-general/src/pii/attachments.rs:
```diff
@@ -272,6 +229,31 @@ impl StringMods for WStr {
+
+    fn fill_content(&mut self, fill_char: char) {
+        let mut buf = [0u8; 1];
+        fill_char.encode_utf8(&mut buf[..]); // this panics for us
```
what do those comments mean?

flub replied:
the trait says we must panic if the char is too wide.  it isn't very obvious from the implementation that we do this, so I point it out.  I'll improve the comments

flub commented on relay-general/src/testutils.rs:
```diff
@@ -27,12 +27,8 @@ macro_rules! assert_eq_bytes_str {
-                    &String::from_utf8_lossy(right),
-                    "\n"
-                )
+                pretty_hex::pretty_hex(left),
+                pretty_hex::pretty_hex(right),
```
This loses the marker of where the differences are btw, my test cases are simple enough that I can just spot things like this so I didn't feel like implementing something better which is probably a lot of work.

untitaker reviewed (approved):
ah now I get how you wanted to share code. I thought you wanted to always match on a `String` at some point

I only reviewed the changes since my last review

jan-auer commented on relay-wstring/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "relay-wstring"
```
Please add this crate to the places where we have to list crates in `relay/src/setup.rs` and `relay/src/main.rs`.

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,512 @@
+//! A UTF-16 little-endian string type.
+//!
+//! The main type in this crate is [WStr] which is a type similar to [str] but with the
```
Since we still need to generate docs with stable, can you please replace all these with proper links and quoting? Currently, it will render like this on our docs page:

![image](https://user-images.githubusercontent.com/1433023/93602116-77606880-f9c2-11ea-97d2-096f0cdf676f.png)

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,512 @@
+    }
+}
+
+impl Default for Utf16Error {
```
Does it make sense to implement `Default`? If we add any field to the error, this default probably has to go.

flub replied:
clippy made me do this...

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,512 @@
+        // Since we always have a valid UTF-16LE string in here we now are sure we always
+        // have a byte at index + 1.  The only invalid thing now is a trailing surrogate.
+
+        let mut buf: [u8; 2] = [0; 2]; // TODO: avoid init
```
Are you still planning to change this? I think we do not need to worry about allocations here.

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,512 @@
+        let mut buf: [u8; 2] = [0; 2]; // TODO: avoid init
+        buf.copy_from_slice(&self.raw[index..index + 2]);
+        let u = u16::from_le_bytes(buf);
+        u & 0xDC00 != 0xDC00
```
You're performing comparisons with `0xDC00` multiple times, sometimes like this and sometimes via `u < 0xDC00`. Can we consolidate, and maybe move this into an `#[inline]` helper function?

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,512 @@
+        u & 0xDC00 != 0xDC00
+    }
+
+    /// Convert to a byte slice.
```
_nit_: Doc comments should be in third person, like the ones above. For example "Converts" rather than "Convert". This also applies to the functions below.

Example: https://doc.rust-lang.org/stable/std/primitive.str.html#method.as_ptr

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,512 @@
+    /// The slice indices are on byte offsets of the underlying UTF-16LE encoded buffer, if
+    /// the subslice is not on character boundaries or otherwise invalid this will return
+    /// `None`.
+    pub fn get<I>(&self, index: I) -> Option<&<I as SliceIndex<WStr>>::Output>
```
Can we mark these as `#[inline]`? Same for all the simple methods.

This can matter since we're crossing crate boundaries.

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -0,0 +1,512 @@
+    for c in std::char::decode_utf16(u16iter) {
+        match c {
+            Ok(_) => (),
+            Err(_) => return Err(Utf16Error::new()),
```
These two can be simplified with a construct similar to (clippy won't let you match):

```rust
if decode_utf16(u16iter).all(|result| result.is_ok()) {
  Ok(())
} else {
  Err(...)
}
```

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -1,7 +1,12 @@
 use regex::bytes::RegexBuilder as BytesRegexBuilder;
-use regex::Regex;
+use regex::{Match, Regex};
+use relay_wstring::{Utf16Error, WStr};
```
_nit_: Please move this import into its own group. We usually separate imports of workspace crates from external crates.

Commit:
A few comment fixes

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -71,43 +65,283 @@ fn apply_regex_to_bytes(
+}
+
+/// Extract the matching encoded slice from the encoded string.
+fn get_wstr_match<'a>(all_text: &str, re_match: Match, all_encoded: &'a mut WStr) -> &'a mut WStr {
```
_nit_: It doesn't really matter, but a nice generic API would be something similar to:

```rust
fn map_str_slice<'a>(source: &str, range: Range<usize>, target: &'a WStr) -> &'a WStr { ... }
```

flub replied:
I prefer this not being generic for now

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -71,43 +65,283 @@ fn apply_regex_to_bytes(
+/// Extract the matching encoded slice from the encoded string.
+fn get_wstr_match<'a>(all_text: &str, re_match: Match, all_encoded: &'a mut WStr) -> &'a mut WStr {
+    let mut encoded_start = 0;
+    let mut encoded_end = 0;
```
I think this needs to be initialized with `all_encoded.len()`. Otherwise, if the end index matches the end of the string, this will remain `0` as the loop below never visits this location.

Commit:
More comment improvements

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -71,43 +65,283 @@ fn apply_regex_to_bytes(
+}
+
+impl<'a> Iterator for MutSegmentIter<'a> {
+    type Item = MutSegment<'a>;
```
This iterator always requires the consumer to perform some unsafe operation afterwards. Three options:

1. Make `MutSegment` an enum that has a `&WStr`variant and do the unsafe stuff internally (where you can verify locally).
2. Give `MutSegmentIter` a type parameter for the encoding, and then implement the `Iterator` trait for specific `Encoding`s. Then you can return a specific item type that contains the `WStr`. I don't like the `Box<RawDecoder>`, but we don't have any option with the `encoding` crate.
3. Always return `WStr` since this is the only application we have for this at the moment.

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -71,43 +65,283 @@ fn apply_regex_to_bytes(
+}
+
+impl<'a> MutSegmentIter<'a> {
+    fn new(data: &'a mut [u8], encoding: impl Encoding) -> Self {
```
_nit_: Please make this a type parameter. There's no benefit to this, but it matches our code style.

flub replied:
do you mean `fn new<E: Encoding>(..., encoding: E)`?

jan-auer replied:
Yes, exactly.

jan-auer reviewed (changes requested):
Thanks, @flub! This is excellent, especially how the string search and UTF-16 handling is separated out into reviewable and maintainable units. Also, I agree with `Iterator` as a choice for string search, even though it requires unsafe internally. 

Reviewed everything but the transmutes in `MutSegmentIter`. See comments below.

Commit:
Wire up docs and logging for new crate

Commit:
Use std::mem::size_of::<u16>() directly

It's a const fn so comes for free and it's more readable this way.

Commit:
Inits are fine apparently

And hard to avoid.

Commit:
Back to fast unsafe code

Commit:
Replace some magic numbers with helper functions

And a lot of doc comments.

Commit:
Doc comments in third person

Commit:
Inline a lot more, taking cues from the stdlib

This inlines a lot more, doing so where the stdlib also does this for str.

Commit:
More elegant iterator usage

Commit:
Correct import blocks

Commit:
Fix full matches

We now have a beautifully failing tests.  It's related but unrelated.
It's interesting.

Commit:
Merge WStr into the SegmentIter

Now we decided to stay with the BytesRegex matching we won't be
parametrising this on the encoding (at least not until we add support
for other encodings).  So it's cleaner for the iterator to handle the
unsafe WStr stuff rather than push it to the user.

Commit:
Avoid the forbidden transmute word

This does the same magic "please interpret these bits over here as the
same type but with a different lifetime" as the transmute.  But with
the same type being forced by using more words and a custom function
instead of a turbofish.  Thus avoiding the word which shall not be
used.

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -6,15 +6,23 @@ use encoding::all::UTF_16LE;
+/// This module extracts encoded strings from within binary blobs, this specifies the
+/// minimum length we require those strings to be before we accept them to match scrubbing
+/// selectors on.
+static MIN_STRING_LEN: usize = 5;
```
_nit_: This can be const.

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -182,13 +187,13 @@ impl StringMods for WStr {
-        let chunks = self.as_bytes_mut().chunks_exact_mut(size);
+        let chunks = self
+            .as_bytes_mut()
+            .chunks_exact_mut(std::mem::size_of::<u16>());
```
_nit_: Since we talked about this offline: If you don't like repeating `size_of`, you could also `const` this on the top of the file. It does read very clunky :)

flub replied:
I *do* like writing the full `std::mem::size_of::<u16>()`!

jan-auer reviewed (approved):
G2G. Two final nit-picks below, aside from the doc comment link situation. Excited about shipping this!

Commit:
I really do want a regex dear clippy

Commit:
Use const instead of static

Commit:
Do not scrub regions twice

If UTF-8 modified any region, exclude that from UTF-16 scrubbing,
since it might now more likely match UTF-16.

Commit:
Merge branch 'master' into feat/utf16-scrub

Commit:
Changelog entry

untitaker commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@
 - Compress outgoing store requests for events and envelopes including attachements using `gzip` content encoding. ([#745](https://github.com/getsentry/relay/pull/745))
 - Relay now buffers all requests until it has authenticated with the upstream. ([#747](//github.com/getsentry/relay/pull/747))
 - Add a configuration option to change content encoding of upstream store requests. The default is `gzip`, and other options are `identity`, `deflate`, or `br`. ([#771](https://github.com/getsentry/relay/pull/771))
+- Add support for scrubbing UTF-16 data in attachments ([#742](https://github.com/getsentry/relay/pull/742))
```
wrong section, create a new `Unreleased` header pls

flub replied:
lol, thanks for catching

Commit:
Fix changelog

oops

Commit:
Merge branch 'master' into feat/utf16-scrub
